### PR TITLE
fix: tab visibility

### LINKED ISSUE #1812: resources inside tabs are not checked against the index? policy
### Context
Consider a has_may relationship, for example, Accounts has many roles and you want to display this in the Account resource.

```ruby
field :roles, as: :has_many
```
Before a simple panel will be visible, avo checks the `index?` policy for roles if you are using pundit or something else. If `index?` is `false`, the panel will not be rendered.

In tab view the pill is rendered and results in a `content missing` message or show loading screen if it is the first item. It would be nice if a resource inside a tab would also be checked against the `index?` policy and the corresponding tab would not be visible.

```ruby
  tabs style: :pills do
    field :roles, as: :has_many
  end
```
```ruby
class RolePolicy < ApplicationPolicy
  def index?
    false
  end
end
```
image

### PR DESCRIPTION
**Pull request by Paul-Bob**
# Description

Fixes #1812

Tab header was wrongly showing buttons to not visible tabs

# Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the [documentation](https://github.com/avo-hq/docs)
- [x] I have added tests that prove my fix is effective or that my feature works

## Manual review steps

1. Step 1
1. Step 2

Manual reviewer: please leave a comment with output from the test if that's the case.

### CODE REVIEW

File: lib/avo/fields/field_extensions/visible_in_different_views.rb [hunk 1]
```diff
@@ -18,7 +18,7 @@ def initialize(id = nil, **args, &block)
         def visible_on?(view)
           raise "No view specified on visibility check." if view.blank?
 
-          send :"show_on_#{view.to_s}"
+          send :"show_on_#{view}"
```
Comment (Paul-Bob) [inline lib/avo/fields/field_extensions/visible_in_different_views.rb:21]: unrelated fix

File: app/components/avo/tab_switcher_component.rb [hunk 2]
```diff
@@ -63,36 +63,31 @@ def selected?(tab)
...
-    tabs.select do |item|
-      visible = true
+    tabs.select do |tab|
+      next false if tab.items.blank?
+      next false if tab.is_field? && !tab.authorized?
 
-      if item.items.blank?
-        visible = false
+      if tab.items.count == 1
+        next false if !single_item_visible?(tab.items.first)
```
Comment (adrianthedev) [inline app/components/avo/tab_switcher_component.rb:71]: can't these be combined into one line?

`tab.items.count == 1 && !single_item_visible?(tab.items.first)`

File: app/components/avo/tab_switcher_component.rb [hunk 3]
```diff
@@ -63,36 +63,31 @@ def selected?(tab)
...
       end
 
-      first_item = item.items.first
-      if item.items.count == 1 && first_item.is_field? && first_item.has_own_panel? && !first_item.visible_on?(view)
-        # Return nil if tab contians a has_many type of fields and it's hidden in current view
-        visible = false
-      end
+      next false if !tab.visible? || !tab.visible_on?(view)
```
Comment (adrianthedev) [inline app/components/avo/tab_switcher_component.rb:74]: let's break these on two lines as they are two separate logical statements.

File: app/components/avo/tab_switcher_component.rb (hunk 2 above)
Comment (adrianthedev) [reply app/components/avo/tab_switcher_component.rb:71]: maybe we coudl add a `has_a_single_item?` method on the tab and make it more readable here?

File: app/components/avo/tab_switcher_component.rb [hunk 4]
```diff
@@ -63,36 +63,31 @@ def selected?(tab)
...
-        visible = item.visible?
-      end
+  private
 
-      if item.respond_to?(:authorized?)
-        visible = item.authorized?
-      end
+  def single_item_visible?(item)
+    # Item is visible if is not a field or don't have its own panel
+    return true if !item.is_field? || !item.has_own_panel?
```
Comment (adrianthedev) [inline app/components/avo/tab_switcher_component.rb:84]: same here break on two lines?

**Review by adrianthedev — APPROVED** [review]
It looks good.
Should we add a test for that?
If the manual tests pass, feel free to merge it.

**Comment by Paul-Bob** [comment]
> It looks good. Should we add a test for that? If the manual tests pass, feel free to merge it.

Yes, we should have a test because this method is kinda complex and sensible, I'll implement a simple one.

**Pull request merged**